kuma/wiki/views/document.py

```diff
@@ -709,70 +697,124 @@ def document(request, document_slug, document_locale):
-
-    contributors = doc.contributors
-    contributors_count = len(contributors)
-    has_contributors = contributors_count > 0
-    other_translations = original_doc.get_other_translations(fields=['title', 'locale', 'slug', 'parent'])
-
-    # Bundle it all up and, finally, return.
-    context = {
-        'document': original_doc,
-        'document_html': doc_html,
-        'toc_html': toc_html,
-        'quick_links_html': quick_links_html,
-        'zone_subnav_html': zone_subnav_html,
-        'body_html': body_html,
-        'contributors': contributors,
-        'contributors_count': contributors_count,
-        'contributors_limit': 6,
-        'has_contributors': has_contributors,
-        'fallback_reason': fallback_reason,
-        'kumascript_errors': ks_errors,
-        'macro_sources': (kumascript.macro_sources(force_lowercase_keys=True)
-                          if ks_errors else
-                          None),
-        'render_raw_fallback': rendering_params['render_raw_fallback'],
-        'seo_summary': seo_summary,
-        'seo_parent_title': seo_parent_title,
-        'share_text': share_text,
-        'search_url': get_search_url_from_referer(request) or '',
-        'analytics_page_revision': doc.current_revision_id,
-        'analytics_en_slug': en_slug,
-        'content_experiment': rendering_params['experiment'],
-        'other_translations': other_translations,
-    }
-    response = render(request, 'wiki/document.html', context)
-    return _set_common_headers(doc, rendering_params['section'], response)
+    def get_etag(*args):
+        return calculate_etag(response.content)
+
+    # The etag decorator not only adds the ETag header to the response,
+    # but also handles conditional responses based on the incoming and
+    # outgoing ETag headers.
+    @etag(get_etag)
+    def get_response(request, *args):
+        return _add_kuma_revision_header(doc, response)
 
+    return get_response(request)
 
-def _document_PUT(request, document_slug, document_locale):
-    """Handle PUT requests as document write API"""
+
+@csrf_exempt
+@require_http_methods(['GET', 'HEAD', 'PUT'])
+@redirect_in_maintenance_mode(methods=['PUT'])
+@allow_CORS_GET
+@accepts_auth_key
+@process_document_path
+@newrelic.agent.function_trace()
+@ratelimit(key='user_or_ip', rate='100/m', block=True)
+def document_api(request, document_slug, document_locale):
+    """
+    View/modify the content of a wiki document, or create a new wiki document.
+    """
+    if request.method == 'PUT':
+        if not (request.authkey and request.user.is_authenticated()):
+            raise PermissionDenied
+        return _document_api_PUT(request, document_slug, document_locale)
+
+    doc = get_object_or_404(
+        Document,
+        slug=document_slug,
+        locale=document_locale
+    )
+
+    section_id = request.GET.get('section', None)
+    response = HttpResponse(doc.get_html(section_id))
```

Comment:
My intention was to avoid changing the semantics of the endpoints as much as possible, but I really like the idea of removing the "section" stuff from the plain document view, and I hate the `$` APIs!